File: selfdoc_core/lints.py
"""The lint-code registry and the verdict rules shared by every check entry point.

``lints.toml`` is the single source of truth for every lint code selfdoc and
selfblog can emit: its severity and its one-line description live there and
nowhere else.  :class:`LintResult` derives ``severity`` from the registry, so
a construction site cannot state a severity and cannot emit an unregistered
code -- both are hard errors rather than conventions.  The lint-code enum in
selfdoc's declared check payload schema (``selfdoc/payload_schemas.py``) and
the lint-rule table in ``docs/check-guide.md`` are pinned to the registry by
tests.


``selfdoc check``, ``selfblog check`` (unified), the post-build lint pass and
the posts-only check all decide the same question -- does this run pass? --
and they used to decide it with four separate copies of the same three
conditions.  One of the copies had drifted: the unified path compared
``documented < total_public`` directly, hardcoding a 100% coverage
requirement, so a project that lowered ``coverage_threshold`` passed
``selfdoc check`` and failed ``selfblog check`` on identical state.

This module owns the rules.  It lives in selfdoc_core because selfblog
depends on selfdoc-core but not on selfdoc: a posts-only selfblog install
must be able to reach the verdict without the selfdoc package present.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from importlib.resources import files

from selfdoc_core import lint_registry_validator as _validator

# Fraction of public symbols that must be documented when a project does not
# say otherwise.  Mirrors the historical CLI default.
DEFAULT_COVERAGE_THRESHOLD = 1.0


# -- The registry -------------------------------------------------------------

_REGISTRY_DOCUMENT = "lints.toml"


class LintRegistryError(RuntimeError):
    """Raised when ``lints.toml`` fails strictspec validation at load.

    A hard error at import time: the lint registry is malformed, so selfdoc
    cannot know what its own lint codes are.  Subclasses ``RuntimeError`` so
    the CLI's existing handlers surface it cleanly.
    """


class LintSuppressionError(RuntimeError):
    """Base for every refusal of a suppression-list entry.

    A suppression list is refused for one of two reasons -- the code is not
    in the registry, or the code is error-severity and therefore not
    suppressible.  Both are the same event to a caller (the list is bad,
    say so and stop), so one ``except`` clause covers them.
    """


class UnknownLintCode(LintSuppressionError):
    """Raised when a lint is constructed with a code the registry does not carry.

    Every emittable code is declared in ``lints.toml``.  An undeclared code
    would carry no severity, would be rejected by the JSON output schema, and
    would be invisible to the documentation table -- so it is refused at the
    construction site instead.
    """


class UnsuppressableLintCode(LintSuppressionError):
    """Raised when a suppression list names an error-severity code.

    Suppression reaches warning-severity codes only.  An error says the
    build is wrong -- a broken emitted reference, a missing description, a
    post whose slug moved -- and silencing it hides the defect rather than
    resolving it, which is exactly how a genuinely broken build once passed
    its own check.  The registry is the severity authority, so the refusal
    is decided there and nowhere else.
    """


@dataclass(slots=True, frozen=True)
class LintSpec:
    """Registry entry for a single lint code."""

    code: str
    severity: str  # "error" or "warning"
    description: str


def _build_registry(raw: bytes) -> dict[str, LintSpec]:
    """Validate raw registry-document bytes and bind them into LintSpecs.

    strictspec is the boundary validator: the document is validated against
    its schema by the generated validator, and only a fully-valid document is
    bound.  Any diagnostic is a hard error (:class:`LintRegistryError`).
    """
    registry, diags = _validator.validate_bytes(raw, "toml")
    if diags:
        detail = "\n".join(f"  {d.path}: {d.message} [{d.code}]" for d in diags)
        raise LintRegistryError(
            f"{_REGISTRY_DOCUMENT} is not a valid lint registry:\n{detail}"
        )
    return {
        entry.code: LintSpec(
            code=entry.code,
            severity=entry.severity,
            description=entry.description,
        )
        for entry in registry.lints
    }


def _load_registry() -> dict[str, LintSpec]:
    """Read, validate, and bind ``lints.toml`` into the runtime registry."""
    raw = files("selfdoc_core").joinpath(_REGISTRY_DOCUMENT).read_bytes()
    return _build_registry(raw)


# Registry order is documentation order: the table renders in this order.
LINT_REGISTRY: dict[str, LintSpec] = _load_registry()


def lint_severity(code: str) -> str:
    """Return the registered severity for *code*, or raise UnknownLintCode."""
    spec = LINT_REGISTRY.get(code)
    if spec is None:
        raise UnknownLintCode(
            f"lint code {code!r} is not in the registry. Every emittable code "
            f"must be declared in selfdoc_core/{_REGISTRY_DOCUMENT} with its "
            f"severity and description."
        )
    return spec.severity


def validate_lint_codes(codes, *, source: str) -> None:
    """Refuse any code in *codes* the registry rejects for suppression.

    Two refusals, both decided by the registry:

    - An unregistered code suppresses nothing and hides the fact that it
      suppresses nothing, so it is a hard error where the list is read
      rather than a silently inert entry.
    - A registered *error*-severity code is not suppressible at all.
      Suppression reaches warnings only; an error means the build is
      wrong, and silencing it hides the defect.

    Args:
        codes: The lint codes to check, in the order the user wrote them.
        source: Where the codes came from, named in the error message
            (e.g. ``"lint_ignore"`` or ``"--ignore"``).

    Raises:
        UnknownLintCode: Naming every unregistered code, not just the first.
        UnsuppressableLintCode: Naming every error-severity code, with its
            severity.  Checked after the unregistered codes, so a typo is
            reported as a typo.
    """
    unknown = [code for code in codes if code not in LINT_REGISTRY]
    if unknown:
        listed = ", ".join(repr(code) for code in unknown)
        raise UnknownLintCode(
            f"{source} names lint code(s) the registry does not carry: "
            f"{listed}. Every suppressible code is declared in selfdoc_core/"
            f"{_REGISTRY_DOCUMENT}; known codes are: "
            f"{', '.join(sorted(LINT_REGISTRY))}."
        )

    errors = [code for code in codes if LINT_REGISTRY[code].severity == "error"]
    if errors:
        listed = ", ".join(
            f"{code} (severity: {LINT_REGISTRY[code].severity})"
            for code in errors
        )
        suppressible = sorted(
            code for code, spec in LINT_REGISTRY.items()
            if spec.severity == "warning"
        )
        raise UnsuppressableLintCode(
            f"{source} names error-severity lint code(s), which cannot be "
            f"suppressed: {listed}. An error says the build is wrong -- fix "
            f"the defect it reports. Suppression reaches warning-severity "
            f"codes only: {', '.join(suppressible)}."
        )


def parse_ignore_codes(raw, *, source: str = "--ignore") -> set[str]:
    """Parse a comma-separated ``--ignore`` value into a validated code set.

    Args:
        raw: The flag value as given, or None/empty when the flag was not
            passed.
        source: Flag name to attribute a rejection to.

    Returns:
        The set of codes to suppress -- empty when nothing was passed.

    Raises:
        UnknownLintCode: When any code is not in the registry.
        UnsuppressableLintCode: When any code is error-severity.
    """
    if not raw:
        return set()
    codes = [part.strip() for part in raw.split(",") if part.strip()]
    validate_lint_codes(codes, source=source)
    return set(codes)


@dataclass(frozen=True)
class LintResult:
    """A single lint diagnostic (e.g. SEO warning).

    ``severity`` is not a constructor argument: it is read from the registry
    for the given code.  That is what keeps severities out of the ~50
    construction sites scattered across the check modules, and what makes an
    unregistered code impossible to emit.

    The instance is frozen: a diagnostic's severity is the registry's answer
    for its code, and nothing downstream may rewrite it after the fact.
    """

    file: str  # relative path within docs/
    line: int | None
    code: str  # e.g. "SEO001"
    message: str
    severity: str = field(init=False)  # derived from the registry

    def __post_init__(self):
        # Frozen instances refuse ordinary assignment; the derived field is
        # written the one way a frozen dataclass permits.
        object.__setattr__(self, "severity", lint_severity(self.code))


def lint_table_rows() -> list[tuple[str, str, str]]:
    """Return ``(code, severity, description)`` for every registered lint.

    In registry (documentation) order.  The documentation's lint-rule table
    is rendered from exactly this.
    """
    return [
        (spec.code, spec.severity, spec.description)
        for spec in LINT_REGISTRY.values()
    ]


def render_lint_table() -> str:
    """Render the registry as the Markdown table the documentation carries."""
    lines = [
        "| Code | Severity | What it checks |",
        "| ---- | -------- | -------------- |",
    ]
    lines.extend(
        f"| {code} | {severity} | {description} |"
        for code, severity, description in lint_table_rows()
    )
    return "\n".join(lines)


def coverage_below_threshold(coverage, config=None) -> bool:
    """True when documented coverage is under the configured threshold.

    Args:
        coverage: A coverage stats object (``total_public``/``documented``),
            or None when the run measured no coverage.
        config: The project configuration dict the run was driven by, read
            for ``coverage_threshold``.  None means "no configuration in
            play", which uses the default threshold.

    Returns:
        False whenever there is nothing to measure -- no coverage object, or
        a project with no public symbols at all.
    """
    if coverage is None or coverage.total_public <= 0:
        return False
    threshold = (config or {}).get(
        "coverage_threshold", DEFAULT_COVERAGE_THRESHOLD,
    )
    return coverage.documented / coverage.total_public < threshold


def check_exit_code(lints, directive_results=(), coverage=None, config=None) -> int:
    """Compute the process exit code for a check run.

    The single definition of "did this check pass": a run fails when a
    directive failed to resolve, when any lint is error-severity, or when
    documented coverage is under the configured threshold.

    Args:
        lints: The lint diagnostics the run produced, already filtered
            through the project's suppression list.
        directive_results: Per-directive resolution results, when the entry
            point resolved directives.  A reduced entry point (the
            post-build lint pass, the posts-only check) passes none.
        coverage: Coverage stats, when the entry point measured coverage.
        config: Project configuration, read for ``coverage_threshold``.

    Returns:
        1 when the run fails, 0 when it passes.
    """
    has_failures = any(dr.status == "FAILED" for dr in directive_results)
    has_errors = any(lint.severity == "error" for lint in lints)
    below = coverage_below_threshold(coverage, config)
    return 1 if (has_failures or has_errors or below) else 0

File: internal/lints/lints.go
// Package lints owns the lint-code registry and the verdict rules every check
// entry point shares.
//
// The embedded lints.toml is the single authority for every lint code selfdoc
// can emit: its severity and its one-line description are declared there and
// nowhere else. [LintResult] derives its severity from the registry, so a
// construction site cannot state a severity and cannot emit an unregistered
// code -- both are refused rather than discouraged. The lint-code enum in the
// declared check payload schema is pinned to the registry by a test, and the
// check guide's lint-rule table is rendered from the registry by the
// "list-lints" directive rather than repeated by hand.
//
// selfdoc's check command, the unified check, the post-build lint pass and the
// posts-only check all decide the same question -- does this run pass? -- and
// they used to decide it with four copies of the same three conditions. One
// copy had drifted: it compared documented against total public symbols
// directly, hardcoding a 100% coverage requirement, so a project that lowered
// its coverage threshold passed one check and failed the other on identical
// state. This package owns the rules.
package lints

import (
	_ "embed"
	"fmt"
	"sort"
	"strings"
	"sync"
)

// DefaultCoverageThreshold is the fraction of public symbols that must be
// documented when a project does not say otherwise. It mirrors the historical
// CLI default.
const DefaultCoverageThreshold = 1.0

// registryDocumentName is the document's name as the error messages spell it.
const registryDocumentName = "lints.toml"

//go:embed lints.toml
var registryDocument []byte

// LintSpec is the registry entry for a single lint code.
type LintSpec struct {
	Code        string
	Severity    string // "error" or "warning"
	Description string
}

// LintRegistryError reports a registry document that failed strictspec
// validation: the lint registry is malformed, so selfdoc cannot know what its
// own lint codes are.
type LintRegistryError struct {
	Message string
}

func (e *LintRegistryError) Error() string { return e.Message }

// Registry is a loaded, validated lint registry.
//
// It keeps the document's order, which is documentation order: the lint-rule
// table renders in it, grouping the families rather than sorting the codes
// alphabetically.
type Registry struct {
	order []string
	specs map[string]LintSpec
}

// Codes returns every registered code in registry (documentation) order.
func (r *Registry) Codes() []string {
	out := make([]string, len(r.order))
	copy(out, r.order)
	return out
}

// Len returns how many codes the registry carries.
func (r *Registry) Len() int { return len(r.order) }

// Spec returns the code's registry entry, reporting whether the registry
// carries it.
func (r *Registry) Spec(code string) (LintSpec, bool) {
	spec, ok := r.specs[code]
	return spec, ok
}

// Has reports whether the registry carries code.
func (r *Registry) Has(code string) bool {
	_, ok := r.specs[code]
	return ok
}

// BuildRegistry validates raw registry-document bytes and binds them into a
// [Registry].
//
// strictspec is the boundary validator: the document is checked against its
// schema by the generated validator, and only a wholly valid document is
// bound. Any diagnostic is a [LintRegistryError].
func BuildRegistry(raw []byte) (*Registry, error) {
	document, diags := ValidateBytes(raw, "toml")
	if len(diags) > 0 {
		var detail strings.Builder
		for _, d := range diags {
			fmt.Fprintf(&detail, "\n  %s: %s [%s]", d.Path, d.Message, d.Code)
		}
		return nil, &LintRegistryError{Message: fmt.Sprintf(
			"%s is not a valid lint registry:%s",
			registryDocumentName, detail.String())}
	}
	reg := &Registry{
		order: make([]string, 0, len(document.Lints)),
		specs: make(map[string]LintSpec, len(document.Lints)),
	}
	for _, entry := range document.Lints {
		reg.order = append(reg.order, entry.Code)
		reg.specs[entry.Code] = LintSpec{
			Code:        entry.Code,
			Severity:    entry.Severity,
			Description: entry.Description,
		}
	}
	return reg, nil
}

// Load reads, validates and binds the embedded registry document.
//
// This is the error-returning door. Production code reads [Registered]
// instead, which loads once.
func Load() (*Registry, error) { return BuildRegistry(registryDocument) }

var (
	registeredOnce sync.Once
	registered     *Registry
	registeredErr  error
)

// Registered returns the shipped lint registry, loading and validating the
// embedded document on first use.
//
// It panics when the document is malformed, which is the Go counterpart of the
// Python surface's import-time crash: the document is embedded in the binary,
// so a diagnostic here means this build of selfdoc does not know what its own
// lint codes are. Use [Load] where an error is wanted.
func Registered() *Registry {
	registeredOnce.Do(func() { registered, registeredErr = Load() })
	if registeredErr != nil {
		panic(registeredErr)
	}
	return registered
}

// ErrLintSuppression is the base every refusal of a suppression-list entry
// wraps.
//
// A suppression list is refused for one of two reasons -- the code is not in
// the registry, or the code is error-severity and therefore not suppressible.
// Both are the same event to a caller (the list is bad, say so and stop), so
// one errors.Is check covers them.
var ErrLintSuppression = fmt.Errorf("lint suppression list refused")

// UnknownLintCodeError reports a code the registry does not carry.
//
// Every emittable code is declared in the registry document. An undeclared
// code would carry no severity, would be rejected by the JSON output schema,
// and would be invisible to the documentation table -- so it is refused at the
// construction site instead.
type UnknownLintCodeError struct {
	Message string
}

func (e *UnknownLintCodeError) Error() string { return e.Message }

// Unwrap reports [ErrLintSuppression], so one errors.Is check covers both
// refusals.
func (e *UnknownLintCodeError) Unwrap() error { return ErrLintSuppression }

// UnsuppressableLintCodeError reports a suppression list that named an
// error-severity code.
//
// Suppression reaches warning-severity codes only. An error says the build is
// wrong -- a broken emitted reference, a missing description, a post whose
// slug moved -- and silencing it hides the defect rather than resolving it,
// which is how a broken build once passed its own check. The registry is the
// severity authority, so the refusal is decided there and nowhere else.
type UnsuppressableLintCodeError struct {
	Message string
}

func (e *UnsuppressableLintCodeError) Error() string { return e.Message }

// Unwrap reports [ErrLintSuppression], so one errors.Is check covers both
// refusals.
func (e *UnsuppressableLintCodeError) Unwrap() error { return ErrLintSuppression }

// LintSeverity returns the registered severity for code, or an
// [UnknownLintCodeError].
func LintSeverity(code string) (string, error) {
	spec, ok := Registered().Spec(code)
	if !ok {
		return "", &UnknownLintCodeError{Message: fmt.Sprintf(
			"lint code '%s' is not in the registry. Every emittable code "+
				"must be declared in the lint registry (internal/lints/%s) "+
				"with its severity and description.", code, registryDocumentName)}
	}
	return spec.Severity, nil
}

// ValidateLintCodes refuses any code the registry rejects for suppression.
//
// Two refusals, both decided by the registry:
//
//   - An unregistered code suppresses nothing and hides the fact that it
//     suppresses nothing, so it is refused where the list is read rather than
//     left silently inert.
//   - A registered error-severity code is not suppressible at all.
//     Suppression reaches warnings only; an error means the build is wrong,
//     and silencing it hides the defect.
//
// source names where the codes came from in the message (for instance
// "lint_ignore" or "--ignore"). Unregistered codes are reported before
// severity, so a typo is reported as a typo, and every offending code is
// named rather than just the first.
func ValidateLintCodes(codes []string, source string) error {
	reg := Registered()

	var unknown []string
	for _, code := range codes {
		if !reg.Has(code) {
			unknown = append(unknown, code)
		}
	}
	if len(unknown) > 0 {
		quoted := make([]string, len(unknown))
		for i, code := range unknown {
			quoted[i] = "'" + code + "'"
		}
		return &UnknownLintCodeError{Message: fmt.Sprintf(
			"%s names lint code(s) the registry does not carry: "+
				"%s. Every suppressible code is declared in the lint "+
				"registry (internal/lints/%s); known codes are: %s.",
			source, strings.Join(quoted, ", "), registryDocumentName,
			strings.Join(sortedCodes(reg), ", "))}
	}

	var errorCodes []string
	for _, code := range codes {
		spec, _ := reg.Spec(code)
		if spec.Severity == "error" {
			errorCodes = append(errorCodes,
				fmt.Sprintf("%s (severity: %s)", code, spec.Severity))
		}
	}
	if len(errorCodes) > 0 {
		var suppressible []string
		for _, code := range sortedCodes(reg) {
			if spec, _ := reg.Spec(code); spec.Severity == "warning" {
				suppressible = append(suppressible, code)
			}
		}
		return &UnsuppressableLintCodeError{Message: fmt.Sprintf(
			"%s names error-severity lint code(s), which cannot be "+
				"suppressed: %s. An error says the build is wrong -- fix "+
				"the defect it reports. Suppression reaches warning-severity "+
				"codes only: %s.",
			source, strings.Join(errorCodes, ", "),
			strings.Join(suppressible, ", "))}
	}

	return nil
}

// sortedCodes returns every registered code sorted by code point, the order
// the refusal messages list them in.
func sortedCodes(reg *Registry) []string {
	out := reg.Codes()
	sort.Strings(out)
	return out
}

// ParseIgnoreCodes parses a comma-separated --ignore value into a validated
// code set.
//
// source names the flag a rejection is attributed to. An empty value means no
// suppression, not an error. Every code is checked through
// [ValidateLintCodes], so an unregistered or error-severity code is refused
// here rather than silently suppressing nothing.
func ParseIgnoreCodes(raw, source string) (map[string]struct{}, error) {
	out := map[string]struct{}{}
	if raw == "" {
		return out, nil
	}
	var codes []string
	for _, part := range strings.Split(raw, ",") {
		if trimmed := strings.TrimSpace(part); trimmed != "" {
			codes = append(codes, trimmed)
		}
	}
	if err := ValidateLintCodes(codes, source); err != nil {
		return nil, err
	}
	for _, code := range codes {
		out[code] = struct{}{}
	}
	return out, nil
}

// LintResult is a single lint diagnostic.
//
// Its severity is not a construction argument: it is read from the registry
// for the given code. That is what keeps severities out of the construction
// sites scattered across the check modules, and what makes an unregistered
// code impossible to emit. Every field is read-only through an accessor,
// because a diagnostic's severity is the registry's answer for its code and
// nothing downstream may rewrite it after the fact.
type LintResult struct {
	file     string
	line     *int
	code     string
	message  string
	severity string
}

// NewLintResult builds a diagnostic for code, deriving its severity from the
// registry.
//
// line is nil for a diagnostic that belongs to a file rather than to one of
// its lines. An unregistered code is an [UnknownLintCodeError].
func NewLintResult(file string, line *int, code, message string) (LintResult, error) {
	severity, err := LintSeverity(code)
	if err != nil {
		return LintResult{}, err
	}
	var lineCopy *int
	if line != nil {
		v := *line
		lineCopy = &v
	}
	return LintResult{
		file:     file,
		line:     lineCopy,
		code:     code,
		message:  message,
		severity: severity,
	}, nil
}

// MustLintResult is [NewLintResult] for a call site whose code is a literal
// declared in the registry document, and panics when it is not.
//
// The panic is the point: an unregistered literal is a defect in this
// repository rather than a condition a run can encounter, and the alternative
// -- an error return threaded through every emission site -- would make the
// emission sites decide what to do about a code that cannot exist.
func MustLintResult(file string, line *int, code, message string) LintResult {
	result, err := NewLintResult(file, line, code, message)
	if err != nil {
		panic(err)
	}
	return result
}

// File returns the diagnostic's path, relative to the docs directory.
func (l LintResult) File() string { return l.file }

// Line returns the 1-based line the diagnostic sits on, or nil when it
// belongs to the file as a whole.
func (l LintResult) Line() *int {
	if l.line == nil {
		return nil
	}
	v := *l.line
	return &v
}

// Code returns the diagnostic's registered lint code.
func (l LintResult) Code() string { return l.code }

// Message returns the diagnostic's human-readable message.
func (l LintResult) Message() string { return l.message }

// Severity returns the registry's severity for this diagnostic's code.
func (l LintResult) Severity() string { return l.severity }

// FilterLints returns the lints whose code is not in ignoreCodes.
//
// An empty or nil ignoreCodes returns the input unchanged, so a run with no
// suppression list does no work.
func FilterLints(lints []LintResult, ignoreCodes map[string]struct{}) []LintResult {
	if len(ignoreCodes) == 0 {
		return lints
	}
	out := make([]LintResult, 0, len(lints))
	for _, lint := range lints {
		if _, ok := ignoreCodes[lint.code]; ok {
			continue
		}
		out = append(out, lint)
	}
	return out
}

// LintTableRow is one row of the rendered lint-rule table.
type LintTableRow struct {
	Code        string
	Severity    string
	Description string
}

// LintTableRows returns a row per registered lint, in registry (documentation)
// order. The documentation's lint-rule table is rendered from this and nothing
// else.
func LintTableRows() []LintTableRow {
	reg := Registered()
	out := make([]LintTableRow, 0, reg.Len())
	for _, code := range reg.Codes() {
		spec, _ := reg.Spec(code)
		out = append(out, LintTableRow{
			Code:        spec.Code,
			Severity:    spec.Severity,
			Description: spec.Description,
		})
	}
	return out
}

// RenderLintTable renders the registry as the Markdown table the documentation
// carries.
func RenderLintTable() string {
	lines := []string{
		"| Code | Severity | What it checks |",
		"| ---- | -------- | -------------- |",
	}
	for _, row := range LintTableRows() {
		lines = append(lines, fmt.Sprintf("| %s | %s | %s |",
			row.Code, row.Severity, row.Description))
	}
	return strings.Join(lines, "\n")
}

// Coverage is what the verdict rules need from a run's coverage measurement:
// how many public symbols there are and how many of them are documented.
//
// The concrete coverage type is the check package's, which measures far more
// than this; declaring the dependency as an interface here keeps this package
// below check in the import graph, as the Python module it ports was (the
// former blog package depended on selfdoc-core but not on selfdoc, so a
// posts-only install must reach the verdict without the check module
// present).
//
// Pass a nil Coverage for a run that measured no coverage. A non-nil
// interface holding a nil pointer is not the same thing and will panic.
type Coverage interface {
	TotalPublic() int
	Documented() int
}

// DirectiveOutcome is what the verdict rules need from one directive's
// resolution result: its status, which is "OK" or "FAILED".
//
// As with [Coverage], the concrete type is the check package's.
type DirectiveOutcome interface {
	Status() string
}

// CoverageBelowThreshold reports whether documented coverage is under the
// configured threshold.
//
// config is the project configuration the run was driven by, read for
// "coverage_threshold"; a nil config means no configuration is in play, which
// uses [DefaultCoverageThreshold]. The answer is false whenever there is
// nothing to measure -- no coverage at all, or a project with no public
// symbols.
func CoverageBelowThreshold(coverage Coverage, config map[string]any) bool {
	if coverage == nil || coverage.TotalPublic() <= 0 {
		return false
	}
	threshold := DefaultCoverageThreshold
	if config != nil {
		if declared, ok := numberValue(config["coverage_threshold"]); ok {
			threshold = declared
		}
	}
	return float64(coverage.Documented())/float64(coverage.TotalPublic()) < threshold
}

// numberValue reads a JSON-decoded configuration value as a float, reporting
// whether it was a number at all. A decoder yields float64, but a
// hand-constructed configuration in a test may carry any integer type, and a
// threshold written as 1 in a config file must mean the same as 1.0.
func numberValue(v any) (float64, bool) {
	switch n := v.(type) {
	case float64:
		return n, true
	case float32:
		return float64(n), true
	case int:
		return float64(n), true
	case int32:
		return float64(n), true
	case int64:
		return float64(n), true
	default:
		return 0, false
	}
}

// CheckExitCode computes the process exit code for a check run.
//
// This is the single definition of "did this check pass": a run fails when a
// directive failed to resolve, when any lint is error-severity, or when
// documented coverage is under the configured threshold. It returns 1 when the
// run fails and 0 when it passes.
//
// lints are the diagnostics the run produced, already filtered through the
// project's suppression list. directiveResults are the per-directive
// resolution results, and are nil for a reduced entry point that resolved no
// directives (the post-build lint pass, the posts-only check). coverage is nil
// for an entry point that measured no coverage.
func CheckExitCode(
	lints []LintResult,
	directiveResults []DirectiveOutcome,
	coverage Coverage,
	config map[string]any,
) int {
	for _, dr := range directiveResults {
		if dr.Status() == "FAILED" {
			return 1
		}
	}
	for _, lint := range lints {
		if lint.severity == "error" {
			return 1
		}
	}
	if CoverageBelowThreshold(coverage, config) {
		return 1
	}
	return 0
}
